### PR TITLE
PAPER 67: Generative Quantum Data Embeddings for Supervised Learning

### PR DESCRIPTION
**Pull request by LF-Vigneux**
Reproducing the  Generative Quantum Data Embeddings for Supervised Learning (https://arxiv.org/abs/2605.30866v1) paper in MerLin. A Claude agent was used for the first version od the reproduction

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,6 +47,7 @@ and any per-file or per-directory notices for details.
...
 | [Photonic Quantum Memristor](papers/qrc_memristor/). Source: [selimovic_2025](https://arxiv.org/abs/2504.18694) | We developed a photonic quantum reservoir computing architecture with the addition of a quantum memristor which acts as a feedback loop (memory). We noticed that in both cases the use of the memristor enhances the non-linear capabilities of the model and this leads to improved performance compared to the non-memristor case and some classical benchmarks. The results are similar to those presented in the paper and the error is 5 times smaller on the learning task of the NARMA dataset.|
+| [Generative Quantum Data Embeddings for Supervised Learning](papers/EGAS/). Source: [Heo_2065](https://arxiv.org/abs/2605.30866v1) | This is a trainable encoding paradigm. It uses a classical GPT model to create iteratively the optimal encoding that is then trained for optimal performance. The training is based on a energy type of problem. There is also a metric introduced that evaluates the potential of the use of quantum to classify the dataset.|
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply README.md]: Done in next commit

File: papers/EGAS/lib/data.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+            rng.choice(idx_neg, per, replace=False),
+        ]
+    )
+    rng.shuffle(sel)
+    X_np, y = X_np[sel], y[sel]
+
+    n_comp = min(n_components, X_np.shape[1])
+    X_std = StandardScaler().fit_transform(X_np)
+    X_pca = PCA(n_components=n_comp, random_state=seed).fit_transform(X_std)
+    X_scaled = MinMaxScaler(feature_range=(0.0, TWO_PI)).fit_transform(X_pca)
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply papers/EGAS/lib/data.py]: NO wassertein uses load_dataset_raw_for_wasserstein

[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply papers/EGAS/lib/data.py]: This is normal the wasserstein uses different loading functions

File: papers/EGAS/run_all_experiments.sh [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+set -e
+
+# Colors for output
+GREEN='\033[0;32m'
+BLUE='\033[0;34m'
+YELLOW='\033[1;33m'
+NC='\033[0m' # No Color
+
+# Configuration
+PYTHON="/Users/lfvigneux/miniconda3/envs/reproduce/bin/python"
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply papers/EGAS/run_all_experiments.sh]: Changed in next commit

File: papers/EGAS/run_all_experiments.sh [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        return
+    fi
+
+    local latest_run=""
+    local latest_mtime=0
+    for run_dir in "$dataset_dir"/run_*; do
+        if [ -d "$run_dir" ]; then
+            local mtime
+            mtime=$(stat -f "%m" "$run_dir" 2>/dev/null || echo 0)
+            if [ "$mtime" -gt "$latest_mtime" ]; then
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply papers/EGAS/run_all_experiments.sh:148]: Changed in next commit

File: papers/EGAS/run_all_experiments.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                    print_success(f"{fullname} results saved")
+            step += len(gate_datasets)
+
+    # Photonic experiments
+    if not args.only_gate:
+        photonic_datasets = [
+            ("MGT", "MAGIC Gamma Telescope"),
+            ("WDGV1", "Waveform DB (multiclass)"),
+        ]
+        for i, (shortname, fullname) in enumerate(photonic_datasets, 1):
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply papers/EGAS/run_all_experiments.py]: CHanged in next commit

File: papers/EGAS/lib/photonic_bias.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
+"""Photonic continuous parameter refinement.
+
+The photonic encoder already gives every data-driven PS gate a trainable phase offset:
+
+    PS(theta_k * r + phi_k)
+
+where ``theta_k`` is fed from ``x[..., data_idx]`` and ``phi_k`` is registered as a MerLin
+trainable parameter.  Refinement therefore trains the encoder's ``phi_k`` parameters directly
+instead of adding a separate bias MLP.
+"""
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply papers/EGAS/lib/photonic_bias.py:10]: Changed in next commit

File: papers/EGAS/lib/photonic_bias.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    seed=0,
+    device="cpu",
+    avg_last=10,
+):
+    """Train encoder phase offsets for a fixed circuit `seq`.
+
+    Returns ``(encoder, E_before, E_after)``.  The returned encoder has its MerLin
+    trainable PS offsets refined in place.  ``hidden`` and ``gain`` are accepted for
+    compatibility with the MLP-bias path and are intentionally unused here.
+    """
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply papers/EGAS/lib/photonic_bias.py]: Changed in next commit

File: papers/EGAS/lib/kernel_svm.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+def classical_svm_accuracy(X_train, y_train, X_test, y_test, kind="linear"):
+    """Compute SVM accuracy on test set. Supports binary and multiclass.
+
+    For multiclass, uses one-vs-rest (OvR) strategy automatically.
+    """
+    scaler = StandardScaler().fit(X_train)
+    Xtr, Xte = scaler.transform(X_train), scaler.transform(X_test)
+    n_classes = len(np.unique(y_train))
+    # Use OvR for multiclass, auto for binary
+    decision_function_shape = "ovr" if n_classes > 2 else "ovr"
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply papers/EGAS/lib/kernel_svm.py]: Changed in next commit

File: papers/EGAS/lib/photonic_kernel_svm.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+def classical_svm_accuracy(X_train, y_train, X_test, y_test, kind="linear"):
+    """Compute SVM accuracy on test set. Supports binary and multiclass.
+
+    For multiclass, uses one-vs-rest (OvR) strategy automatically.
+    """
+    scaler = StandardScaler().fit(X_train)
+    Xtr, Xte = scaler.transform(X_train), scaler.transform(X_test)
+    n_classes = len(np.unique(y_train))
+    # Use OvR for multiclass, auto for binary
+    decision_function_shape = "ovr" if n_classes > 2 else "ovr"
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply papers/EGAS/lib/photonic_kernel_svm.py]: Changed in next commit

File: papers/EGAS/utils/run_fig3_repeats.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            "dataset": DATASET,
+            "repeats": REPEATS,
+            "G_mean": Gm,
+            "G_std": Gs,
+            "B_mean": Bm,
+            "B_std": Bs,
+        },
+        open(out / "fig3_data.json", "w"),
+        indent=2,
+    )
```
[the remark this answers is not in the document]
Comment (LF-Vigneux) [reply papers/EGAS/utils/run_fig3_repeats.py]: Changed in next commit

File: papers/EGAS/README.md [hunk 11]
```diff
@@ -0,0 +1,289 @@
+# Generative Quantum Data Embeddings for Supervised Learning — Reproduction
+
+Reduced-compute reproduction of **EGAS** (Energy-based Generative Architecture Search) and its
+Wasserstein-geometry diagnostic, with a MerLin photonic counterpart.
+
+## Reference and Attribution
+- **Paper:** J. Heo and D. K. Park, *Generative Quantum Data Embeddings for Supervised Learning*,
+  [arXiv:2605.30866v1](https://arxiv.org/abs/2605.30866v1) (29 May 2026), quant-ph / cs.LG. Yonsei University.
+- No official code repository was found; this is an independent reimplementation from the paper
```
Comment (CassNot) [inline papers/EGAS/README.md]: There is ! https://github.com/qDNA-yonsei/Generative-QDE

File: papers/EGAS/lib/data.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+        [
+            rng.choice(idx_pos, per, replace=False),
+            rng.choice(idx_neg, per, replace=False),
+        ]
+    )
+    rng.shuffle(sel)
+    X_np, y = X_np[sel], y[sel]
+
+    n_comp = min(n_components, X_np.shape[1])
+    X_std = StandardScaler().fit_transform(X_np)
```
Comment (CassNot) [inline papers/EGAS/lib/data.py]: The original repo (https://github.com/qDNA-yonsei/Generative-QDE) does not use `StandardScaler` and apply PCA on raw features

File: papers/EGAS/lib/data.py (hunk 2 above)
Comment (CassNot) [inline papers/EGAS/lib/data.py]: In the original code, the angle rescaling to 0, 2pi is per-sample (so row-wise) so each row's own min/max mapped to [0,2π]
Here, it is per-feature (so column-wise) MinMaxScaler (which should lead to different encoding)

File: papers/EGAS/lib/egas.py
Comment (CassNot) [inline papers/EGAS/lib/egas.py:1]: Here, your CNOT and MultiRZ are restricted to a nearest-neighbour ring but in the original work, they are enumerated over all qubit pairs (with control different than target)

File: papers/EGAS/lib/gpt.py
Comment (CassNot) [inline papers/EGAS/lib/gpt.py:1]: Original (https://github.com/qDNA-yonsei/Generative-QDE/blob/main/EGAS_and_refinement/model.py): a full nanoGPT-style transformer (model.py reminds me of https://github.com/karpathy/nanogpt), tied embedding/output weights, flash attention, fused AdamW. I would say this is not trivial and not really mentioned in the paper.

Here: your TokenGPT is a minimal `nn.TransformerEncoder`-based model + the configs actually are different so your GPT looks smaller

File: papers/EGAS/lib/data.py (hunk 2 above)
Comment (CassNot) [reply papers/EGAS/lib/data.py]: The authors do it manually here https://github.com/qDNA-yonsei/Generative-QDE/blob/main/EGAS_and_refinement/data.py

File: papers/EGAS/lib/photonic_circuits.py [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+    computation_space: ml.ComputationSpace = ml.ComputationSpace.UNBUNCHED,
+):
+    circuit, input_parameters, trainable_parameters = create_perceval_circuit(
+        sequence, n_modes=n_modes
+    )
+
+    ps_data_indices = [data_idx for gate, _, data_idx, _ in sequence if gate == "PS"]
+    ps_r_factors = [r for gate, _, _, r in sequence if gate == "PS"]
+
+    class BiasMLP(nn.Module):
```
Comment (CassNot) [inline papers/EGAS/lib/photonic_circuits.py:85]: Caution with initialization here: zero-initialized final MLP layer+ a small gain=10.0 + Tanh hidden layers could lead to very small gradients at initialization
This could lead to 100–400 epochs at lr=5e-4 not being enough

File: papers/EGAS/tests/test_photonic_impl.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    y_train = [0, 1]
+    X_test = [[1.0, 0.0], [0.0, 1.0]]
+    y_test = [0, 1]
+
+    acc = qksvm_accuracy(model, X_train, y_train, X_test, y_test)
+
+    assert acc == 1.0
+
+
+def test_refine_bias_returns_same_energy_when_no_trainable_parameters():
```
Comment (CassNot) [inline papers/EGAS/tests/test_photonic_impl.py:78]: But there is no test to assert that `refine_bias` actually moves the energy on a circuit that has trainable PS tokens
All tests could pass with the bias mechanism working or not

Comment (CassNot) [reply papers/EGAS/tests/test_photonic_impl.py:78]: (because there is 0 PS token)

**Review by CassNot — CHANGES_REQUESTED** [review]
Nice work and non-trivial as well !
I found the original repo ! https://github.com/qDNA-yonsei/Generative-QDE so I am understanding a bit more the paper in light of the code.
My analysis is therefore based on the cross-checking of both repo :) I also asked Claude to help me cross-review your 2 repo on the gate-based settings (to gain some time in the comparison of all these files) so the comments here and below are a mix of my findings and Claude's (but approved by me)

One difference worth noting is that in the orignal repo, there is no StandardScaler applied before PCA and PCA is ran on raw inputs
I prefer your train/test construction with non-overlapping slices with repeats per dataset

Question
- how did you chose the token pool for the photonic version ?

**Comment by LF-Vigneux** [comment]
For the photonic pool, I tried to be as close to the gate-based reproduction as possible. Indeed, the PS can be trainable or fixed with fixed PS values. For the two qubit gates, I only have the choice of doing nearest neighbor beam splitters. That is why I added two types since the ratio of one-mode gates vs entangling gates was off.

File: papers/EGAS/README.md (hunk 11 above)
Comment (LF-Vigneux) [reply papers/EGAS/README.md]: Added it

File: papers/EGAS/lib/data.py (hunk 12 above)
Comment (LF-Vigneux) [reply papers/EGAS/lib/data.py]: Changed in next commit

File: papers/EGAS/lib/data.py (hunk 2 above)
Comment (LF-Vigneux) [reply papers/EGAS/lib/data.py]: I am just transposing before and after the scaling

File: papers/EGAS/lib/egas.py
Comment (LF-Vigneux) [reply papers/EGAS/lib/egas.py:1]: Changed in next commit

File: papers/EGAS/lib/photonic_circuits.py (hunk 13 above)
Comment (LF-Vigneux) [reply papers/EGAS/lib/photonic_circuits.py:85]: Changed in next commit

File: papers/EGAS/data/generative_quantum_embeddings/DB_X.csv
Comment (CassNot) [inline papers/EGAS/data/generative_quantum_embeddings/DB_X.csv:1]: I would move all /data in the appropriated /data folder that contains all relevant data for the papers (root)
This comment is for all files in data/generative_quantum_embeddings

File: papers/EGAS/notebook.ipynb
Comment (CassNot) [inline papers/EGAS/notebook.ipynb:1]: I believe this notebook is lacking some explanations and context. It should be self-explainable. I would advise adding the original paper image + maybe explain again (really quickly) what the paper does

**Review by CassNot — CHANGES_REQUESTED** [review]
Nice changes !

- 2 comments on `data/` files and notebook explanations
- `ruff` needs to be applied 

With that done, it should be good to go !

File: papers/EGAS/data/generative_quantum_embeddings/DB_X.csv
Comment (LF-Vigneux) [reply papers/EGAS/data/generative_quantum_embeddings/DB_X.csv:1]: Changed it!

Comment (CassNot) [reply papers/EGAS/data/generative_quantum_embeddings/DB_X.csv:1]: They are still here under papers/EGAS/data

**Review by CassNot — CHANGES_REQUESTED** [review]
Last changes remaining (not addressed) the data/ folder is still in papers/EGAS/data instead of data/ in the root folder

**Comment by LF-Vigneux** [comment]
It is now done, I forgot to remove it manually, it was in my gitignore

**Comment by CassNot** [comment]
@LF-Vigneux but there are no new folder in the /root/data. Is it expected ? (I thought you would just git move it)

**Comment by LF-Vigneux** [comment]
Updated the general gitignore 😅

**Review by CassNot — APPROVED** [review]
